### ISSUE TITLE
AssertionError while using TrailingStrategy

### ISSUE THREAD

**Issue by zlpatel** [kernc/backtesting.py#316]
I am seeing this error when try to run backtest with 10000 days worth data. It doesn't happen for all the stocks, but some. e.g AAPL, AMZN.

```
Index: 1946  Close: 35.88999938964844  ATR: 1.6433353276970901
Index: 1947  Close: 35.779998779296875  ATR: 1.5673829350797757
Index: 1948  Close: 35.779998779296875  ATR: 1.5232842084983242
Index: 1949  Close: 36.029998779296875  ATR: 1.4951925556138177
Index: 1950  Close: 36.13999938964844  ATR: 1.438393141851363
Index: 1951  Close: 35.65999984741211  ATR: 1.3906505347952947
Traceback (most recent call last):
  File "backtesting_usa.py", line 116, in <module>
    output = bt.run()
  File "/usr/local/lib/python3.8/site-packages/backtesting/backtesting.py", line 1165, in run
    strategy.next()
  File "backtesting_usa.py", line 83, in next
    super().next()
  File "/usr/local/lib/python3.8/site-packages/backtesting/lib.py", line 446, in next
    trade.sl = max(trade.sl or -np.inf,
  File "/usr/local/lib/python3.8/site-packages/backtesting/backtesting.py", line 633, in sl
    self.__set_contingent('sl', price)
  File "/usr/local/lib/python3.8/site-packages/backtesting/backtesting.py", line 652, in __set_contingent
    assert price is None or 0 < price < np.inf
AssertionError
```

I tried printing the index values for each iteration and I found that for this dataset, it processed till index 1951 and failed on 1952.

If I reduce the backtesting period to 5000, it works because it doesn't come across that piece of data (index 1952). I checked that data and verified the close price, and calculated ATR(14). 

Below is the code snippet:

```py
class Breakout(TrailingStrategy):
	timeperiod = 10
	position_size_decimal = 0.2
	min_close = []
	max_close = []
	def init(self):
		super().init()
		self.ema20 = self.I(EMA,self.data.Close,20,overlay=True)
		self.atr14 = self.I(ATR,self.data.High,self.data.Low,self.data.Close,14)
		self.set_atr_periods(14)
		self.set_trailing_sl(1)
		print(type(self.data.Close))
		self.min_close, self.max_close = MINMAX(self.data.Close, timeperiod=self.timeperiod)

	def next(self):
		super().next()
		index = len(self.data)-1
		print('Index: '+f'{index} '+' Close: '+f'{self.data.Close[index]} '+' ATR: '+f'{self.atr14[index]}')
		if not self.position.is_long and self.min_close[index] > (self.max_close[index] * 0.98) and self.max_close[index] < (self.min_close[index] * 1.02):
			#print('BUY ZONE:'+ f'{self.data.index[index]} {self.rsi2[index]} {self.rsi5[index]} {self.ema200[index]} {self.data.Close[index]}')
			self.buy(size=self.position_size_decimal)
```

I download data using yfinance python api.  Any help will be greatly appreciated.

_Originally posted by @zlpatel in https://github.com/kernc/backtesting.py/discussions/315_

**Comment by kernc** [comment]
These are the culprit lines in `TrailingStrategy`:
https://github.com/kernc/backtesting.py/blob/0b2325f94ff5dbe3486202a08cd981a3cd123d41/backtesting/lib.py#L446-L447
The way `max()` can produce a value for which `assert price is None or 0 < price < np.inf` fails is if `trade.sl` is unset (you don't pass initial `self.buy(..., sl=)`) _and_ `self.data.Close[-1] - self.__atr[-1] * self.__n_atr` is _negative_.
Meaning, the ATR in $ is greater than the price in $.

Can you propose a fix?

**Comment by zlpatel** [comment]
@kernc I looked at that code earlier. Still not sure how this code would return negative?
 `self.data.Close[-1] - self.__atr[-1] * self.__n_atr` 
In this case, 
...
Index: 1951  Close: 35.65999984741211  ATR: 1.3906505347952947
...
so 35.66 - (1.39*1) = 34.27. which is a positive value. Am I missing something here?

**Comment by kernc** [comment]
:thinking: Ideally, someone would fire up a debugger and place a breakpoint [here](https://github.com/kernc/backtesting.py/blob/0b2325f94ff5dbe3486202a08cd981a3cd123d41/backtesting/lib.py#L446-L447), then see what's what, and why the assertion fails.
Can you provide a piece of your data to try on?

**Comment by zlpatel** [comment]
@kernc Apologies. I am not quite familiar with python debugging. reading about debugging using pdb. But here's the dataset that I am using for my test. I will try to debug it.
[AMZN.csv](https://github.com/kernc/backtesting.py/files/6380971/AMZN.csv)

**Comment by zlpatel** [comment]
@kernc so I started the debugger. price value is coming out as -70.88 and hence the AssertionError. I am still checking the price calculation.

**Comment by kernc** [comment]
You can always inject a little helper code:
```py
if price < 0:  # our special condition of interest
    import ipdb; ipdb.set_trace()
```
Then you can do an `up`, `up` to go from the assertion up the stack to the `TrailingStrategy.next()` where all the fun is.

**Comment by zlpatel** [comment]
@kernc  I have found the culprit. It's `self.__atr[-1]`

`
/usr/local/lib/python3.8/site-packages/backtesting/lib.py(433)set_atr_periods()
-> self.__atr = atr
(Pdb) pp atr[-1]
62.03802490234375
`
`
/usr/local/lib/python3.8/site-packages/backtesting/lib.py(440)set_trailing_sl()
-> self.__n_atr = n_atr
(Pdb) pp n_atr
1.5
`
`
> /usr/local/lib/python3.8/site-packages/backtesting/lib.py(446)next()
-> trade.sl = max(trade.sl or -np.inf,
(Pdb) pp self.data.Close[-1]
22.170000076293945
`

so per the formula: `trade.sl = max(trade.sl or -np.inf, 
                self.data.Close[-1] - self.__atr[-1] * self.__n_atr)`

22.17 - (62.04*1.5) = 70.89

the code shouldn't be using atr[-1]. It should be using atr[index] where index = len(self.data)-1

**Comment by zlpatel** [comment]
@kernc do you want me to make changes and submit a merge request?

**Comment by kernc** [comment]
Excellent work!

Your proposed fix sounds good. If you'd like to submit a PR, that would be splendid! Don't forget a small, contrived unit/regression test.

**Comment by zlpatel** [comment]
@kernc thanks. do I need to have write permission to this repo?

**Comment by zlpatel** [comment]
@kernc  oh nevermind. it says it will create a new branch.

**Comment by zlpatel** [comment]
@kernc I ran into this with some other data, but this time it's not an issue with the SL calculation :) I put a breakpoint at the same place and found that for some reason, it has the same value for ATR. Below are the values I printed from code. It fails on Index 84.  Notice atr values for past 2 indexes as well. 

>Index: 82 Open: 0.4375 High: 0.5 Low: 0.375 Close: 0.4375
>Index: 82 atr: 0.20375
>Index: 83 Open: 0.5 High: 0.5 Low: 0.4375 Close: 0.5
>Index: 83 atr: 0.20375
>Index: 84 Open: 0.4375 High: 0.5 Low: 0.375 Close: 0.375
>Index: 84 atr: 0.20375

Attaching data here if you want to test it.
[NVR.csv](https://github.com/kernc/backtesting.py/files/6408624/NVR.csv)
